miden/src/cli/debug.rs

```diff
@@ -0,0 +1,219 @@
+use super::data::{InputFile, ProgramFile};
+use crate::{
+    repl::{print_mem_address, print_stack},
+    VmState,
+};
+use miden::VmStateIterator;
+use rustyline::Editor;
+use std::path::PathBuf;
+use structopt::StructOpt;
+
+#[derive(StructOpt, Debug)]
+#[structopt(name = "Debug", about = "Debug a miden program")]
+pub struct DebugCmd {
+    /// Path to .masm assembly file
+    #[structopt(short = "a", long = "assembly", parse(from_os_str))]
+    assembly_file: PathBuf,
+    /// Path to input file
+    #[structopt(short = "i", long = "input", parse(from_os_str))]
+    input_file: Option<PathBuf>,
+}
+
+impl DebugCmd {
+    pub fn execute(&self) -> Result<(), String> {
+        println!("============================================================");
+        println!("Debug program");
+        println!("============================================================");
+
+        // load program from file and compile
+        let program = ProgramFile::read(&self.assembly_file)?;
+
+        let program_hash: [u8; 32] = program.hash().into();
+        println!("Debugging program with hash {}... ", hex::encode(program_hash));
+
+        // load input data from file
+        let input_data = InputFile::read(&self.input_file, &self.assembly_file)?;
+
+        // fetch the stack and program inputs from the arguments
+        let stack_inputs = input_data.parse_stack_inputs()?;
+        let advice_provider = input_data.parse_advice_provider()?;
+
+        // generate debug trace
+        let mut vm_state_iter = processor::execute_iter(&program, stack_inputs, advice_provider);
+        let mut vm_state = vm_state_iter
+            .next()
+            .ok_or(format!("failed to start iterating vm state"))?
+            .unwrap();
+
+        // initializing readline
+        let mut rl = Editor::<()>::new().expect("Readline couldn't be initialized");
```

Comment:
It would be nice to have `emacs` or `vim` mode enabled by default:

https://docs.rs/rustyline/latest/rustyline/config/struct.Builder.html#method.edit_mode

Also history:

https://docs.rs/rustyline/latest/rustyline/config/struct.Builder.html#method.auto_add_history

Reply:
I've added vim bindings behind a `-vi` flag.  Have fun :smiley: 
`./target/debug/miden debug -vi -a ../examples/examples/shamir-secret-share.masm`